tools/replica/reorder_remd_traj.py

```diff
@@ -0,0 +1,571 @@
+    
+    :param enefn: ascii file (readable by numpy.loadtxt) containing an array
+                  u[r,n] of *total* potential energy for the n-th frame for
+                  the r-th replica.
+    
+    :param frametuple_dict: dict containing a tuple (replica #, frame #) 
+                            for each temp.
+    
+    :param temps: array of temps. used in the REMD simulation
+    
+    :param nprod: number of production timesteps. Last (nprod / writefreq) 
+                  frames from the end will be written to disk.
+    
+    :param writefreq: traj dump frequency in LAMMPS
+    
+    :param kB : Boltzmann constant to set the energy scale.
+                Default is in kcal/mol
+    
+    Returns: logw: dict, logw[l][k,n] gives the log weights from the
+                   n-th frame of the k-th temp. *ordered* trajectory
+                   to reweight to the l-th temp.
+    
+    """
+    
+    try:
+        import pymbar
+    except ImportError:
+        print("""
+              Configurational log-weight calculation requires pymbar.
+              Here are some options to install it:
+              conda install -c omnia pymbar
+              pip install pymbar
+          
+              To install the dev. version directly from github, use:
+              pip install pip install git+https://github.com/choderalab/pymbar.git
+              """)
+              
+    u_rn = np.loadtxt(enefn)
+    ntemps = u_rn.shape[0] # number of temps.
+    nframes = int(nprod / writefreq) # number of frames at each temp.
+    
+    # reorder the temps
+    u_kn = np.zeros([ntemps, nframes], float)
+    for k in range(ntemps):
+        frame_tuple = frametuple_dict[k][-nframes:]
+        for i, (rep, frame) in enumerate(frame_tuple):
+            u_kn[k, i] = u_rn[rep, frame]
+    
+    # prep input for pymbar    
+    #1) array of frames at each temp.
+    nframes_k = nframes * np.ones(ntemps, np.uint8)
+    
+    #2) inverse temps. for chosen energy scale
+    beta_k = 1.0 / (kB * temps)
+    
+    #3) get reduced energies (*ONLY FOR THE CANONICAL ENSEMBLE*)
+    u_kln = np.zeros([ntemps, ntemps, nframes], float)
+    for k in range(ntemps):
+        for l in range(ntemps):
+            u_kln[ k, l, 0:nframes_k[k] ] = beta_k[l] * u_kn[k, 0:nframes_k[k]]
+    
+    # run pymbar and extract the free energies
+    print("\nRunning pymbar...")
+    mbar = pymbar.mbar.MBAR(u_kln, nframes_k, verbose = True)
+    f_k = mbar.f_k
+
+    # calculate the log-weights
+    print("\nExtracting log-weights...")
+    log_nframes = np.log(nframes)
+    logw = dict( (k, np.zeros([ntemps, nframes], float)) for k in range(ntemps) )
+    for l in range(ntemps):
+        # get log-weights to reweight to this temp.
+        for k in range(ntemps):
+            for n in range(nframes):
+                num = -beta_k[k] * u_kn[k,n]
+                denom = f_k - beta_k[k] * u_kn[k,n]
+                logw[l][k,n] = num - logsumexp(denom) - log_nframes
```

Comment:
My feeling is that if the order of the loops is changed the execution will be faster. It seems that the expression num - logsumexp(denom) - log_nframes is independent of l. If the outer l-loop becomes the inner one then things should be faster. Also some vectorization can be done on either k or n variable.
for k in range(ntemps):
  num = -beta_k[k] * u_kn[k,:]
  denom = f_k - beta_k[k] * u_kn[k,:]
  for l in range(ntemps):
    logw[l][k,:] = num - logsumexp(denom) - log_nframes

Reply:
Shifted the loop with index l to the innermost position. However f_k is actually a 1x ntemps size array so that this vectorization will not work. This step is actually super fast so in my opinion, there's little benefit in producing more convoluted code (while trying to vectorize this).